test/the-rewarder/README.md

```diff
@@ -0,0 +1,587 @@
+### Loading rewards
+In the setup process, the original test internally parses 1000 records in **JSON** format and uploads them to the **distrubutor** contract. However, there is a problem: Halmos does not support the required cheat codes, namely `vm.projectRoot()`, `vm.readFile()` and `vm.parseJson()`. We will work around this problem in a somewhat dirty but effective way. Instead of parsing the **JSON**, we will immediately explicitly insert the bytes into the right place. 
+
+First, let's log the necessary bytes from the original **TheRewarder.t.sol**:
+```solidity
+function _loadRewards(string memory path) private view returns (bytes32[] memory leaves) {
+    console.logBytes(vm.parseJson(vm.readFile(string.concat(vm.projectRoot(), path))));
+    ...
+```
+```javascript
+$ forge test --mp test/the-rewarder/TheRewarder.t.sol -vvv
+...
+Logs:
+  0x000...e962
+  0x000...b3d4
+```
+Insert it into the Halmos test:
+```solidity
+function setUp() public {
+    ...
+    bytes32[] memory dvtLeaves = _loadRewardsDVT();
+    bytes32[] memory wethLeaves = _loadRewardsWETH();
+    ...
+}
+...
+function _loadRewardsDVT() private view returns (bytes32[] memory leaves) {
+    Reward[] memory rewards =
+        abi.decode(hex"000...e962", (Reward[]));
+    assertEq(rewards.length, BENEFICIARIES_AMOUNT);
+
+    leaves = new bytes32[](BENEFICIARIES_AMOUNT);
+    for (uint256 i = 0; i < BENEFICIARIES_AMOUNT; i++) {
+        leaves[i] = keccak256(abi.encodePacked(rewards[i].beneficiary, rewards[i].amount));
+    }
+}
+    
+    
+function _loadRewardsWETH() private view returns (bytes32[] memory leaves) {
+    Reward[] memory rewards =
+        abi.decode(hex"000...b3d4", (Reward[]));
+    assertEq(rewards.length, BENEFICIARIES_AMOUNT);
+
+    leaves = new bytes32[](BENEFICIARIES_AMOUNT);
+    for (uint256 i = 0; i < BENEFICIARIES_AMOUNT; i++) {
+        leaves[i] = keccak256(abi.encodePacked(rewards[i].beneficiary, rewards[i].amount));
+    }
+}
+```
+This thing works for a long time in Halmos: it took all of 1 minute on my machine to create **dvt** and **weth** leaves.
+### Dealing with Merkle functions
+Again cryptography puts a spanner in our works. This time when we try to run the test we get an error:
+```javascript
+$ halmos --solver-timeout-assertion 0 --function check_theRewarder
+...
+Error: setUp() failed: HalmosException: No successful path found in setUp()
+```
+The problem is in `merkle.getRoot()`:
+```solidity
+function getRoot(bytes32[] memory data) public pure virtual returns (bytes32) {
+    require(data.length > 1, "won't generate root for single leaf");
+    while (data.length > 1) {
+        data = hashLevel(data);
+    }
+    return data[0];
+}
+```
+Halmos doesn't do well with large loops. The good news is that we don't have to look for root in the runtime. It is enough to calculate it once, even in the original forge test and hardcode it:
```

Comment:
nice workaround

Reply:
although I can sense there will be issues with just passing a hash value (without halmos knowing a hash of what)

Reply:
is the loop really the cause of the failed `setUp()`? the error message is saying that all paths explored in setUp revert

Reply:
Described more precisely